### PR TITLE
perf: improved pool balancing

### PR DESCRIPTION
**Pull request by ronag**
If all clients are busy then put requests in a queue and dispatch on first client that is no longer busy.

### CODE REVIEW

File: lib/client.js [hunk 1]
```diff
@@ -794,6 +798,10 @@ function resume (client) {
     }
 
     if (!client.pending) {
+      if (client.connected && !client[kWriting] && !client[kReset]) {
+        // TODO: Only emit if was busy.
+        client.emit('drain')
```
Comment (ronag) [inline lib/client.js:803]: I'm a little unsure about the naming here... Also it might be emitted multiple times, though that is probably mostly harmless.

Comment (mcollina) [reply lib/client.js:799]: drain is fine!

File: lib/pool.js [hunk 2]
```diff
@@ -17,49 +21,104 @@ class Pool {
...
     }
 
+    this[kQueue] = []
+    this[kPendingIdx] = 0
     this[kClients] = Array.from({
       length: connections || 10
     }, () => new Client(url, options))
+
+    for (const client of this[kClients]) {
+      client.on('drain', () => {
```
Comment (mcollina) [inline lib/pool.js:31]: can you please create this closure in a separate function? closures and loops are.. weird.

Comment (ronag) [reply lib/pool.js:31]: it needs to capture the client instance though?

Comment (mcollina) [reply lib/pool.js:31]: you don't need it, it is `this` in the function (if you remove the arrow).

Comment (mcollina) [reply lib/pool.js:31]: Yes, something like `client.on('drain', onDrain(this))`

Comment (ronag) [reply lib/pool.js]: How about now?

**Comment by ronag** [comment]
> It would be good to have a test to cover this.

I will revisit this later and write more tests.

**Pull request merged**